crates/exex/src/lib.rs

```diff
@@ -0,0 +1,158 @@
+    /// that Reth is allowed to prune them.
+    ///
+    /// On reorgs, it's possible for the height to go down.
+    pub finished_height: watch::Receiver<BlockNumber>,
+}
+
+/// The main trait for the custom ExEx.
+///
+/// The ExEx should listen to the canonical state notifications and process them, updating the
+/// `finished_height` in [ExExHandle] accordingly.
+pub trait IntoExEx<DB: Database, P: FullProvider<DB>> {
+    /// Spawns the ExEx.
+    ///
+    /// The ExEx should listen to the canonical state notifications and process them, updating the
+    /// `finished_height` in [ExExHandle] accordingly. The ExEx may receive the same notification
+    /// again, and ExEx's reaction to that should be idempotent. Example cases when a duplicate
+    /// notification is possible are:
+    /// 1. Reth sends a notification during the pipeline sync, but restarts before committing the
+    /// data to the database.
+    /// 2. The ExEx doesn't update the `finished_height` in the handle.
+    ///
+    /// Returns a future that should be polled to advance the ExEx and a handle to the ExEx.
+    fn spawn_exex(
+        self,
+        provider: P,
+        canon_state_notifications: CanonStateNotifications,
+    ) -> (BoxFuture<'static, ()>, ExExHandle);
+}
+
+/// The meta information about the ExEx.
+#[derive(Debug, Clone)]
+pub struct ExExMeta {
+    /// The identifier of the ExEx. It will appear in logs and metrics.
+    pub id: String,
+    /// The height from which the ExEx should start processing blocks. Earlier blocks will not be
+    /// sent to the ExEx.
+    ///
+    /// In case when the ExEx needs all blocks from the genesis, it should be set to `0`.
+    pub from_height: BlockNumber,
+}
+
+/// The managed version of the [IntoExEx] trait that covers the common use case and simplifies the
+/// usage.
+pub trait ExEx<DB: Database, P: FullProvider<DB>>: Send + Sync + Unpin + 'static {
+    /// Initializes the ExEx. Called once when the ExEx is spawned. ExEx should persist the
+    /// dependencies such as the provider to re-use them in the future when processing state
+    /// notifications.
+    ///
+    /// Returns the meta information about the ExEx.
+    fn init(&self, provider: P) -> ExExMeta;
+
+    /// Processes the state notification. Once processed, Reth is allowed to prune the blocks from
+    /// notification.
+    ///
+    /// The ExEx may receive the same notification again, and ExEx's reaction to that should be
+    /// idempotent. For example, if Reth sends a notification during the pipeline sync, but
+    /// restarts before committing the data to the database, the same notification will be sent
+    /// again.
+    ///
+    /// [Result::Err] does not stop the ExEx, and is used only for observability purposes. The ExEx
+    /// will continue to receive new notifications.
+    fn on_state_notification(
+        &self,
+        notification: CanonStateNotification,
+    ) -> impl Future<Output = Result<(), ExExError>> + Send;
```

Comment:
imo all of this should be handled be the exex task internally

Reply:
thats what `IntoExEx` is